android/src/main/java/com/tailscale/ipn/App.kt

```diff
@@ -0,0 +1,478 @@
+        const val STATUS_NOTIFICATION_ID = 1
+        const val NOTIFY_CHANNEL_ID = "tailscale-notify"
+        const val NOTIFY_NOTIFICATION_ID = 2
+        private const val PEER_TAG = "peer"
+        private const val FILE_CHANNEL_ID = "tailscale-files"
+        private const val FILE_NOTIFICATION_ID = 3
+        private val mainHandler = Handler(Looper.getMainLooper())
+        lateinit var appInstance: App
+        private fun isEmpty(str: String?) = str.isNullOrEmpty()
+        fun startActivityForResult(act: Activity, intent: Intent?, request: Int) {
+            val f: Fragment = act.getFragmentManager().findFragmentByTag(PEER_TAG)
+            f.startActivityForResult(intent, request)
+        }
+
+        fun getApplication(): App {
+            return appInstance
+        }
+    }
+
+    val dns = DnsConfig()
+    var autoConnect = false
+    var vpnReady = false
+    private lateinit var connectivityManager: ConnectivityManager
+    private lateinit var app: libtailscale.Application
+
+    override fun getPlatformDNSConfig(): String = dns.dnsConfigAsString
+
+    override fun isPlayVersion(): Boolean = MaybeGoogle.isGoogle()
+
+    override fun log(s: String, s1: String) {
+        Log.d(s, s1)
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        val dataDir = this.filesDir.absolutePath
+        app = Libtailscale.start(dataDir, this)
+        Request.setApp(app)
+        Notifier.setApp(app)
+        connectivityManager = this.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        setAndRegisterNetworkCallbacks()
+        createNotificationChannel(NOTIFY_CHANNEL_ID, "Notifications", NotificationManagerCompat.IMPORTANCE_DEFAULT)
+        createNotificationChannel(STATUS_CHANNEL_ID, "VPN Status", NotificationManagerCompat.IMPORTANCE_LOW)
+        createNotificationChannel(FILE_CHANNEL_ID, "File transfers", NotificationManagerCompat.IMPORTANCE_DEFAULT)
+        appInstance = this
+        applicationScope.launch {
+            Notifier.tileReady.collect { isTileReady ->
+                setTileReady(isTileReady)
+            }
+        }
+    }
+
+    override fun onTerminate() {
+        super.onTerminate()
+        Notifier.stop()
+        applicationScope.cancel()
+    }
+
+    fun connect() {
+        val callback: (Result<Ipn.Prefs>) -> Unit = { result ->
+            if (result.isSuccess) {
+                val prefs = result.getOrNull()
+                prefs?.let {
+                    Log.d("App", "Connect: preferences updated successfully: $prefs")
+                } ?: Log.d("App", "Connect: No prefs available, but call was successful.")
+            } else {
+                val error = result.exceptionOrNull()
+                Log.e("App", "Failed to update preferences: ${error?.message}")
+            }
+        }
+        setWantRunning(true, callback)
+    }
+
+    fun disconnect() {
+        val callback: (Result<Ipn.Prefs>) -> Unit = { result ->
```

Comment:
Nit.. If all we're going to do is log, we can probably just define a function somewhere that does it.  I don't think you need to log the success case either.